src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -178,6 +178,17 @@ trait Infer extends Checkable {
     def context: Context
     import InferErrorGen._
 
+    // flag used to turn off tupling when alternatives of an overload include a Java-defined method
+    private[this] var tupleForApplicability = true
+    def withoutTuplingForApplicability[A](body: => A): A =
```

Comment:
This seems like a new place to store this sort of state. Usually I would expect it to be a `ContextMode` or a `Mode` bit. I'm not sure that storing this here is necessarily incorrect, but I can't really be sure.

The class of bug that worries me is when, in the apparent seclusion of `try body finally ...`, a type completer is triggered that bears witness the state you're toggling.